docs: roll out guide (#829)
kubernetes-sigs/gateway-api-inference-extension: site-src/guides/rollout.md

# Rollout

The goal of this guide is to show you how to perform incremental roll out operations, 
which gradually deploy new versions of your inference infrastructure. 
You can update nodes, base models, and LoRA adapters with minimal service disruption. 
This page also provides guidance on traffic splitting and rollbacks to help ensure reliable deployments.

The following use cases are supported:

*   [LoRA adapter update roll out](#lora-adapter-rollout)
*   [Node (compute, accelerator) update roll out](#node-update-rollout)
*   [Base model update roll out](#base-model-rollout)

> ## **Prerequisites**

Reviewer: ```suggestion
## Prerequisites
```

Follow the steps in the [main guide](index.md)

> ## **lora-adapter-rollout** 

Reviewer: ```suggestion
## LoRA Adapter Rollout
```

> LoRA adapter update roll outs let you deploy new versions of fine-tuned models in phases, 

Reviewer: ```suggestion
LoRA adapter rollouts let you deploy new versions of LoRA adapters in phases, 
```

without altering the underlying base model or infrastructure. 

> Use LoRA adapter update roll outs to test improvements, bug fixes, or new features in your LoRA adapters.

Reviewer: ```suggestion
Use LoRA adapter rollouts to test improvements, bug fixes, or new features in your LoRA adapters.
```

"max_tokens": 100,
"temperature": 0
}'
```

### Finish the rollout

Modify the InferenceModel to direct 100% of the traffic to the latest version of the adapter.

```yaml
apiVersion: inference.networking.x-k8s.io/v1alpha2
kind: InferenceModel
metadata:
  name: food-review
spec:
  modelName: food-review
  criticality: Standard
  poolRef:
    name: vllm-llama3-8b-instruct
  targetModels:
  - name: food-review-2
    weight: 100
```

Unload the older versions from the servers by updating the LoRA syncer ConfigMap to list the older version under the `ensureNotExist` list:

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: vllm-llama3-8b-instruct-adapters
data:
  configmap.yaml: |
    vLLMLoRAConfig:
      name: vllm-llama3-8b-instruct-adapters
      port: 8000
      defaultBaseModel: meta-llama/Llama-3.1-8B-Instruct
      ensureExist:
        models:
        - id: food-review-2
          source: Kawon/llama3.1-food-finetune_v14_r8
      ensureNotExist:
        models:
        - id: food-review-1
          source: Kawon/llama3.1-food-finetune_v14_r8
```

With this, all requests should be served by the new adapter version.

> ## **node-update-rollout**

Reviewer: ```suggestion
## InferencePool rollout
```

> Node update roll outs safely migrate inference workloads to new node hardware or accelerator configurations. 

Reviewer: Node rollout is confusing because it can mean updating the kubernetes node versions. Pls update the doc and be specific that this is about rolling out a new InferencePool. 

While upgrading nodes can be a reason to do an InferencePool rollout, there can be other reasons (trying a new accerlerator, a new model server , etc.)

This process happens in a controlled manner without interrupting model service. 
Use node update roll outs to minimize service disruption during hardware upgrades, driver updates, or security issue resolution.

1.  **Create a new `InferencePool`**: deploy an `InferencePool` configured with the
    updated node or hardware specifications.

> 1.  **Split traffic using an `HTTPRoute`**: configure an `HTTPRoute` to distribute

Reviewer: Pls provide an `HTTPRoute` example yaml

    traffic between the existing and new `InferencePool` resources. Use the `weight`
    field in `backendRefs` to manage the traffic percentage directed to the new
    nodes.
1.  **Maintain a consistent `InferenceModel`**: retain the existing
    `InferenceModel` configuration to ensure uniform model behavior across both
    node configurations.

> 1.  **Retain original resources**: keep the original `InferencePool` and nodes

Reviewer: Can you update the steps similar to the LoRA rollout guide above, and provide example yamls for each step, such as:
1. Prepare the new InferencePool and InferenceModels, and EPP
1. Apply traffic split
1. Verify 
1. Finish the rollout 
1. Delete old InferencePool, InferenceModels and EPP

    active during the roll out to enable rollbacks if needed.

For example, you can create a new `InferencePool` named `llm-new`. Configure
this pool with the same model configuration as your existing `llm`
`InferencePool`. Deploy the pool on a new set of nodes within your cluster. Use
an `HTTPRoute` object to split traffic between the original `llm` and the new
`llm-new` `InferencePool`. This technique lets you incrementally update your
model nodes.

See an example here: [replace-inference-pool](replacing-inference-pool.md)

> ## **base-model-rollout**

Reviewer: This seems similar to the InferencePool rollout above. If so, can we remove this one, and add a list of use cases when an InferencePool rollout is required, e.g., updating base model, updating major model server version (or try a different model server), updating accelerator type

Base model updates roll out in phases to a new base LLM, retaining compatibility
with existing LoRA adapters. You can use base model update roll outs to upgrade to
improved model architectures or to address model-specific issues.

1.  **Deploy new infrastructure**: Create new nodes and a new `InferencePool`
    configured with the new base model that you chose.
1.  **Configure traffic distribution**: Use an `HTTPRoute` to split traffic
    between the existing `InferencePool` (which uses the old base model) and the new
    `InferencePool` (using the new base model). The `backendRefs weight` field
    controls the traffic percentage allocated to each pool.
1.  **Maintain `InferenceModel` integrity**: keep your `InferenceModel`
    configuration unchanged. This ensures that the system applies the same LoRA
    adapters consistently across both base model versions.
1.  **Preserve rollback capability**: retain the original nodes and
    `InferencePool` during the roll out to facilitate a rollback if necessary.

> See an example here: [replace-inference-pool](replacing-inference-pool.md)

Reviewer 2: it looks like https://github.com/kubernetes-sigs/gateway-api-inference-extension/blob/1f62b02a3a3f90094b9132eaa0116e25d73a0c72/site-src/api-types/inferencepool.md?plain=1#L62 also refers to this file, do we want to redirect that to here first instead? Not strictly necessary, just calling out the multi-link. Or perhaps that page can refer back to this rollout guide if needed
